cloudify_cli/replace_certificates_config.py

```diff
@@ -0,0 +1,377 @@
+                    and not stderr.channel.recv_stderr_ready() \
+                    and not stdout.channel.recv_ready():
+                stdout.channel.shutdown_read()
+                stdout.channel.close()
+                break
+
+        stdout.close()
+        stderr.close()
+
+        exit_status = stdout.channel.recv_exit_status()
+        if exit_status != 0:
+            raise CloudifyCliError(output)
+
+        return output
+
+    @abc.abstractmethod
+    def _create_paths_mapping(self):
+        pass
+
+    def run_command(self, command):
+        try:
+            self.logger.debug('Running `{0}` on '
+                              '{1}'.format(command, self.host_ip))
+            return self._blocking_exec_command(command)
+        except CloudifyCliError as err:
+            self._command_failed(command, err.message)
+
+    @contextmanager
+    def sftp_connection(self):
+        sftp = None
+        try:
+            sftp = self.client.open_sftp()
+            yield sftp
+        finally:
+            if sftp:
+                sftp.close()
+
+    def put_file(self, local_path, remote_path):
+        self.logger.debug('Copying {0} to {1} on host {2}'.format(
+            local_path, remote_path, self.host_ip))
+        with self.sftp_connection() as sftp:
+            sftp.put(local_path, remote_path)
+
+    def _command_failed(self, command, err_msg):
+        self.errors_list.append('The command `{0}` on host {1} '
+                                'failed with error: '
+                                '{2}'.format(command, self.host_ip, err_msg))
+
+    def replace_certificates(self):
+        self.logger.info(('Replacing certificates on host '
+                          '{0}'.format(self.host_ip)))
+        self.run_command('cfy_manager replace-certificates -i '
+                         '{0}'.format(REMOTE_CONFIG_FILE_PATH))
+
+    def validate_certificates(self):
+        self.logger.info(('Validating certificates on host '
+                          '{0}'.format(self.host_ip)))
+        self._pass_certificates()
+        self.run_command('cfy_manager replace-certificates -i {0}'
+                         ' --only-validate'.format(REMOTE_CONFIG_FILE_PATH))
+
+    def needs_to_replace_certificates(self):
+        return any(self.certs_mapping.values())
+
+    def _pass_certificates(self):
+        self._prepare_new_certs_dir()
+        for cert_name, new_cert_path in self.certs_mapping.items():
+            if new_cert_path:
+                remote_cert_path = CERTS_MAPPING[cert_name]
+                self.certs_dict[cert_name] = remote_cert_path
+                self.put_file(new_cert_path, remote_cert_path)
+
+        self._pass_config_file()
+
+    def _pass_config_file(self):
+        local_path = LOCAL_CONFIG_FILE_PATH.format(self.host_ip + '_')
+        with open(local_path, 'w') as config_file:
+            yaml.dump(self.certs_dict, config_file)
+
+        self.put_file(local_path, REMOTE_CONFIG_FILE_PATH)
```

Comment:
fabric would let you put a stringio! :P

Reply:
Nice! I tested it now and it seems that StringIO doesn't except dictionaries as inputs?

Reply:
@YoniItzhak stringio is a file-like object, just in memory instead of on disk (so you dont need to remove it etc). You still write to it, same as with a file, so `yaml.dump(xxx, some_stringio)` should work

Reply:
@tehasdf Thanks for the explanation! I'm not using the file anymore so it's fine :)